pkg/distributor/distributor.go

```diff
@@ -1223,12 +1226,27 @@ func (d *Distributor) prePushValidationMiddleware(next PushFunc) PushFunc {
 			skipLabelCountValidation := d.cfg.SkipLabelCountValidation || req.GetSkipLabelCountValidation()
 
 			// Note that validateSeries may drop some data in ts.
+			rawSamples := len(ts.Samples)
+			rawHistograms := len(ts.Histograms)
 			shouldRemove, validationErr := d.validateSeries(now, &req.Timeseries[tsIdx], userID, group, skipLabelValidation, skipLabelCountValidation, minExemplarTS, maxExemplarTS)
 
 			if countDroppedNativeHistograms {
 				droppedNativeHistograms += len(ts.Histograms)
 			}
 
+			dedupedSamplesAndHistograms := (rawSamples - len(ts.Samples)) + (rawHistograms - len(ts.Histograms))
+			if dedupedSamplesAndHistograms > 0 {
+				for _, l := range ts.Labels {
+					if l.Name != labels.MetricName {
+						continue
+					}
+					if _, exists := dedupedPerMetric[l.Value]; exists || len(dedupedPerMetric) < maxTrace {
```

Comment:
The `len` check is cheaper than the `exists` check so should come first.
Even better: it could go on the `if` outside the loop, and then the loop can be replaced with `UnsafeMetricNameFromLabelAdapters`.

Reply:
Can you please check my revision in this regard.